torch/nn/modules/module.py

```diff
@@ -382,6 +382,151 @@ def add_module(self, name: str, module: Optional['Module']) -> None:
+                (net_b): Module(
+                    (net_c): Module(
+                        (conv): Conv2d(16, 33, kernel_size=(3, 3), stride=(2, 2))
+                    )
+                    (linear): Linear(in_features=100, out_features=200, bias=True)
+                )
+            )
+
+        (The diagram shows an ``nn.Module`` ``A``. ``A`` has a nested 
+        submodule ``net_b``, which itself has two submodules ``net_c`` 
+        and ``linear``. ``net_c`` then has a submodule ``conv``.)
+
+        To check whether or not we have the ``linear`` submodule, we
+        would call ``get_submodule("net_b.linear")``. To check whether
+        we have the ``conv`` submodule, we would call
+        ``get_submodule("net_b.net_c.conv")``.
+
+        The runtime of ``get_submodule`` is bounded by the degree
+        of module nesting in ``target``. A query against 
+        ``named_modules`` achieves the same result, but it is O(N) in 
+        the number of transitive modules. So, for a simple check to see
+        if some submodule exists, ``get_submodule`` should always be
+        used.
+
+        Args:
+            target: The fully-qualified string name of the submodule
+                to look for. (See above example for how to specify a
+                fully-qualified string.)
+
+        Returns:
+            Optional[torch.nn.Module]: The submodule referenced by
+                ``target`` if it exists, otherwise ``None``. ``None``
+                will also be returned if the target string resolves
+                to something that is not an ``nn.Module``.
+        """
+        atoms: List[str] = target.split(".")
+        mod: torch.nn.Module = self
+
+        for item in atoms:
+
+            if not hasattr(mod, item):
+                return None
+
+            mod = getattr(mod, item)
+
+            if not isinstance(mod, torch.nn.Module):
+                return None
+
+        return mod
+
+    def get_parameter(self, target: str) -> Optional["Parameter"]:
+        """
+        Returns the parameter given by ``target`` if it exists,
+        otherwise returns ``None``.
+
+        See the docstring for ``get_submodule`` for a more detailed
+        explanation of this method's functionality as well as how to
+        correctly specify ``target``.
+
+        Args:
+            target: The fully-qualified string name of the Parameter
+                to look for. (See ``get_submodule`` for how to specify a
+                fully-qualified string.)
+
+        Returns:
+            Optional[torch.nn.Parameter]: The Parameter referenced by
+                ``target`` if it exists, otherwise ``None``. ``None``
+                will also be returned if the target string resolves
+                to something that is not an ``nn.Parameter``.
+        """
+        atoms: List[str] = target.split(".")
+        module_atoms, param_name = atoms[:-1], atoms[-1]
+        mod: torch.nn.Module = self
+
+        for item in module_atoms:
```

Comment:
This code is common to all 3 new functions here, can it be factored?

Reply:
Good point, thanks!